Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/TapToPlace/TapToPlace.cs

```diff
@@ -0,0 +1,467 @@
+                {
+                    rotateAccordingToSurface = value;
+                }
+            }
+        }
+
+        private const int IgnoreRaycastLayer = 2;
+
+        private const int DefaultLayer = 0;
+
+        // The current ray is based on the TrackedTargetType (Controller Ray, Head, Hand Joint)
+        protected RayStep currentRay;
+
+        protected RaycastHit currentHit;
+
+        protected float previousFrameNumber = 0;
+
+        protected override void Start()
+        {
+            // Solver base class
+            base.Start();
+
+            // If tap to place is added via script, set the GameObjectToPlace as this gameobject 
+            if (GameObjectToPlace == null)
+            {
+                GameObjectToPlace = gameObject;
+            }
+
+            if(!ColliderPresent)
+            {
+                Debug.LogError("The GameObjectToPlace does not have a collider attached, please attach a collider");
+            }
+
+            SurfaceNormalOffset = GetComponent<Collider>().bounds.extents.z;
+
+            // What if the spatial awareness is not enabled?
+            spatialAwarenessSystemAcecess = CoreServices.SpatialAwarenessSystem as IMixedRealityDataProviderAccess;
+
+            // Get the spatial mesh display option on start
+            // I do not know if this is legal, check on this
+            initialSpatialMeshDisplayOption = InitialSpatialMeshDisplayOption;
+
+            if (AutoStart)
+            {
+                StartPlacement();
+            }
+            else
+            {
+                SolverHandler.UpdateSolvers = false;
+            }
+        }
+
+        private void StartPlacement()
+        {
+            CoreServices.InputSystem?.RegisterHandler<IMixedRealityPointerHandler>(this);
+
+            // Temporarily change the game object layer to IgnoreRaycastLayer to enable a surface hit beyond the game object
+            gameObject.layer = IgnoreRaycastLayer;
+
+            if (SpatialMeshVisible)
+            {
+                ToggleSpatialMeshVisibility(true);
+            }
+
+            SolverHandler.UpdateSolvers = true;
+
+            IsBeingPlaced = true;
+        }
+
+        private void StopPlacement()
+        {
+            // Change the game object layer back to default which enables raycast hits again
+            gameObject.layer = DefaultLayer;
+
+            if (SpatialMeshVisible)
+            {
+                ToggleSpatialMeshVisibility(false);
+            }
+
+            SolverHandler.UpdateSolvers = false;
+
+            IsBeingPlaced = false;
+
+            CoreServices.InputSystem?.UnregisterHandler<IMixedRealityPointerHandler>(this);
+        }
+
+        /// <inheritdoc/>
+        public override void SolverUpdate()
+        {
+            // Make sure the Transform target is not null, added for the case where auto start is true 
+            // and the tracked target type is the controller ray, if the hand is not in the frame we cannot
```

Comment:
Given this comment, would it be worth scoping this conditional to just that tracked target type, or does it not really matter if we do this check for other other target types? Just trying to understand the comment more.

Reply:
`if (SolverHandler.TransformTarget != null)` was added for the case where `AutoStart` is true and the `TrackedTargetType` is not the head (i.e. Controller Ray or Hand Joint).  The head gaze ray is always present on application start, so there is not a case where it could be null.  

Unlike the head gaze ray,  the controller ray or hand joints as a `TrackedTargetType` come in and out of tracking, which create potential for a null `TransformTarget` on Start.   For example, if we have an object with `AutoStart` enabled and the TrackedTargetType is controller ray,  but the controller ray is not tracking yet on Start, then we are trying to calculate where to place an object based on a null position.